File: trace/analys/systrace/analys.cc
#include <boost/regex.hpp>
#include <chrono>
#include <filesystem>
#include <format>
#include <fstream>
#include <map>
#include <print>

enum class Status {
    OK,
    ERR,
};

class Elapse {
public:
    explicit Elapse(const std::string& name);
    explicit Elapse(std::nullptr_t) = delete;
    ~Elapse();

protected:
    Elapse(const Elapse&) = delete;
    Elapse(Elapse&&) = delete;
    Elapse& operator=(const Elapse&) = delete;
    Elapse& operator=(Elapse&&) = delete;

private:
    std::string _name {};
    std::chrono::high_resolution_clock::time_point _start {};
    std::chrono::high_resolution_clock::time_point _end {};
};

Elapse::Elapse(const std::string& name)
    : _name(name)
{
    _start = std::chrono::high_resolution_clock::now();
}

Elapse::~Elapse()
{
    _end = std::chrono::high_resolution_clock::now();
    std::println("{} cost time {}", _name, std::chrono::duration<double, std::milli>(_end - _start));
}

class ITraceItem {
public:
    virtual ~ITraceItem() = default;

protected:
    ITraceItem() = default;

public:
    std::string task {};
    long pid { -1 };
    long tgid { -1 };
    std::string cpu {};
    std::string delay {};
    double timestamp { 0.0 };
    std::string function {};
    std::string message {};
};

class TraceItem : public ITraceItem {
public:
    TraceItem() = default;
    virtual ~TraceItem() = default;
};

static const std::string str(R"(^ +(.+)\-(\d+) +\( *(\S+)\) +\[(\d+)\] +(\S+) +(\d+\.\d+): +([a-zA-Z0-9_]+): *(.*)$)");

auto preprocess(const std::string& path, std::map<unsigned long long, TraceItem>& trace) -> Status
{
    Elapse elapse(std::string(__func__));
    std::ifstream ifs(path);
    std::string line {};
    boost::regex pattern(str);
    boost::smatch matched_part {};
    auto status { Status::ERR };
    if (ifs.is_open()) {
        unsigned long long index { 0 };
        while (std::getline(ifs, line)) {
            bool found = boost::regex_search(line, matched_part, pattern);
            if (found) {
                TraceItem item;
                std::string task = matched_part[1];
                std::string pid = matched_part[2];
                std::string tgid = matched_part[3];
                std::string cpu = matched_part[4];
                std::string delay = matched_part[5];
                std::string timestamp = matched_part[6];
                std::string function = matched_part[7];
                std::string message = matched_part[8];

                item.task = task;
                try {
                    item.pid = std::stol(pid);
                } catch (...) {
                    std::println("[{}][{}][{}][{}][{}][{}][{}][{}]", task, pid, tgid, cpu, delay, timestamp, function, message);
                    break;
                }
                try {
                    item.tgid = tgid.compare("-----") ? std::stol(tgid) : -1;
                } catch (...) {
                    std::println("[{}][{}][{}][{}][{}][{}][{}][{}]", task, pid, tgid, cpu, delay, timestamp, function, message);
                    break;
                }
                item.cpu = cpu;
                item.delay = delay;
                item.timestamp = std::stod(timestamp);
                item.function = function;
                item.message = message;
                trace[index++] = item;
            } else {
                std::println("{}", line);
            }
        }
        status = Status::OK;
    }

    return status;
}

auto print_trace(const std::map<unsigned long long, TraceItem>& trace) -> Status
{
    Elapse elapse(std::string(__func__));
    auto status { Status::ERR };
    std::for_each(trace.cbegin(), trace.cend(), [](auto &e){ std::format("[{}][{}][{}][{}][{}][{}][{}][{}]", e.second.task, e.second.pid, e.second.tgid, e.second.cpu, e.second.delay, e.second.timestamp, e.second.function, e.second.message);});
    // std::for_each(trace.cbegin(), trace.cend(), [](auto &e){ std::println("[{}][{}][{}][{}][{}][{}][{}][{}]", e.second.task, e.second.pid, e.second.tgid, e.second.cpu, e.second.delay, e.second.timestamp, e.second.function, e.second.message);});
    status = Status::OK;

    return status;
}

auto main(int argc, char* argv[]) -> int
{
    std::map<unsigned long long, TraceItem> trace {};
    auto status { Status::ERR };
    if (!argv[1])
    {
        std::println("{}: {}", __LINE__, "please input a file path as arg!");
        return -1;
    }
    std::string perfetto_path(argv[1]);
    std::filesystem::path perfetto(perfetto_path);
    if (!std::filesystem::exists(perfetto))
    {
        std::println("{}: {} {}", __LINE__, perfetto_path, "not exist!");
        return -1;
    }
    if (!std::filesystem::is_regular_file(perfetto))
    {
        std::println("{}: {} {}", __LINE__, perfetto_path, "is not a regular file!");
        return -1;
    }
    std::string stem = perfetto.stem().string();
    std::string systrace_path = stem + ".html";
    std::filesystem::path systrace(systrace_path);
    if (!std::filesystem::exists(systrace))
    {
        std::println("{}: {} {}", __LINE__, systrace_path, "systrace not exist!");
        std::string cmd("./traceconv systrace " + perfetto_path + " " + systrace_path);
        system(cmd.c_str());
    }
    if (!std::filesystem::is_regular_file(systrace))
    {
        std::println("{}: {} {}", __LINE__, perfetto_path, "is not a regular file!");
        return -1;
    }

    status = preprocess(systrace_path, trace);
    if (Status::ERR == status) {
        return -1;
    }
    status = print_trace(trace);
    if (Status::ERR == status) {
        return -1;
    }

    return 0;
}

File: android/perfetto/analys/systrace/analys.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

'''
pip3 install perfetto
py -3 analys.py -f redmi.perfetto-trace
'''

import logging
import os
import re
import sys
import time
from getopt import getopt
from pathlib import Path
from pprint import pprint

pattern = r"^ +(.+)\-(\d+) +\( *(\S+)\) +\[(\d+)\] +(\S+) +(\d+\.\d+): +([a-zA-Z0-9_]+): *(.*)$"

class Elapse:
    def __init__(self, name):
        self.name = name
        self.start = time.perf_counter()
        print(self.name, "start...")

    def __del__(self):
        self.end = time.perf_counter()
        print(self.name, "process cost time", self.end - self.start)

def preprocess(filename):
    elapse = Elapse(sys._getframe().f_code.co_name)
    trace = {}
    with open(filename, 'r') as file:
        i = 0
        for line in file:
            match = re.search(pattern, line)
            if match:
                task = match.group(1)
                pid = match.group(2)
                tgid = match.group(3)
                cpu = match.group(4)
                delay = match.group(5)
                timestamp = match.group(6)
                function = match.group(7)
                message = match.group(8)
                pid = int(pid)
                if '-----' == tgid:
                    tgid = -1
                trace[i] = {'task':task,
                            'pid':pid,
                            'tgid':tgid,
                            'cpu':cpu,
                            'delay':delay,
                            'timestamp':timestamp,
                            'function':function,
                            'message':message}
                i = i + 1
            else:
                pprint(line)
    return trace

def print_trace(trace):
    elapse = Elapse(sys._getframe().f_code.co_name)
    for key, value in trace.items():
        line = "[{}][{}][{}][{}][{}][{}][{}][{}][{}]".format(
            key,
            value['task'],
            value['pid'],
            value['tgid'],
            value['cpu'],
            value['delay'],
            value['timestamp'],
            value['function'],
            value['message'])
        print(line)

def main():
    elapse = Elapse(sys._getframe().f_code.co_name)
    opt, filenames = getopt(sys.argv[1:], 'f', ['file'])
    if not filenames:
        logging.error("please input a filename!")
        return
    for perfetto in filenames:
        logging.debug("input file: " + perfetto)
        perfetto_path = Path(perfetto)
        if not perfetto_path.exists():
            logging.error(perfetto + ": not exist!")
            return
        if not perfetto_path.is_file():
            logging.error(perfetto + ": not regular file!")
            return
        stem = perfetto_path.stem
        systrace = stem + ".html"
        systrace_path = Path(systrace)
        if not systrace_path.exists():
            logging.info(systrace + ": not exist!")
            cmd = "py -3 traceconv systrace {} {}".format(perfetto, systrace)
            print(cmd)
            os.system(cmd)
        trace = preprocess(systrace)
        print_trace(trace)

if __name__ == '__main__':
    main()